pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -2,35 +2,194 @@ package oauthclient
+// Based on the namespace and names provided, it builds a map of resource name to redirect URIs.
+// The redirect URIs represent the default values as specified by the resource.
+// These values can be overridden by user specified data.
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
+
+var emptyGroupKind = unversioned.GroupKind{} // Used with static redirect URIs
+var routeGroupKind = routeapi.SchemeGroupVersion.WithKind(RouteKind).GroupKind()
+
+// TODO add ingress support
+// var ingressGroupKind = routeapi.SchemeGroupVersion.WithKind(IngressKind).GroupKind()
+
 type saOAuthClientAdapter struct {
 	saClient     kclient.ServiceAccountsNamespacer
 	secretClient kclient.SecretsNamespacer
+	routeClient  osclient.RoutesNamespacer
+	// TODO add ingress support
+	//ingressClient ??
 
 	delegate    oauthclient.Getter
 	grantMethod oauthapi.GrantHandlerType
+
+	decoder runtime.Decoder
+}
+
+// model holds fields that could be used to build redirect URI(s).
+// The resource components define where to get the default redirect data from.
+// If specified, the uri components are used to override the default data.
+// As long as the resulting URI(s) have a scheme and a host, they are considered valid.
+type model struct {
+	scheme string
+	port   string
+	path   string
+	host   string
+
+	group string
+	kind  string
+	name  string
+}
+
+// getGroupKind is used to determine if a group and kind combination is supported.
+func (m *model) getGroupKind() unversioned.GroupKind {
+	return unversioned.GroupKind{Group: m.group, Kind: m.kind}
+}
+
+// updateFromURI updates the data in the model with the user provided URL data.
+func (m *model) updateFromURI(u *url.URL) {
+	m.scheme, m.host, m.path = u.Scheme, u.Host, u.Path
+	if h, p, err := net.SplitHostPort(m.host); err == nil {
+		m.host = h
+		m.port = p
+	}
+}
+
+// updateFromReference updates the data in the model with the user provided object reference data.
+func (m *model) updateFromReference(r *oauthapi.RedirectReference) {
+	m.group, m.kind, m.name = r.Group, r.Kind, r.Name
+}
+
+type modelList []model
+
+// getNames determines the unique, non-empty resource names specified by the models.
+func (ml modelList) getNames() sets.String {
+	data := sets.NewString()
+	for _, model := range ml {
+		if len(model.name) > 0 {
+			data.Insert(model.name)
+		}
+	}
+	return data
+}
+
+// getRedirectURIs uses the mapping provided by a namesToObjMapperFunc to enumerate all of the redirect URIs
+// based on the name of each resource.  The user provided data in the model overrides the data in the mapping.
+// The returned redirect URIs may contain duplicate and invalid entries.
+func (ml modelList) getRedirectURIs(objMapper map[string]redirectURIList) redirectURIList {
+	var data redirectURIList
+	for _, m := range ml {
+		if uris, ok := objMapper[m.name]; ok {
+			for _, uri := range uris {
+				u := uri
+				u.merge(&m)
```

Comment:
will this mutate the item held by objMapper? testcase like this would be good:

```
serviceaccounts.openshift.io/oauth-redirecturi.1=https://:myport
serviceaccounts.openshift.io/oauth-redirectref.1=<ref to route foo>
serviceaccounts.openshift.io/oauth-redirecturi.2=https:///mypath
serviceaccounts.openshift.io/oauth-redirectref.2=<ref to route foo>
```

make sure we get back two distinct redirect uris, one with a custom port, one with a custom path